velox/functions/sparksql/aggregates/tests/ModeAggregateTest.cpp

```diff
@@ -0,0 +1,508 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/Aggregate.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+#include "velox/functions/lib/aggregates/tests/utils/AggregationTestBase.h"
+#include "velox/functions/sparksql/aggregates/Register.h"
+
+using namespace facebook::velox::exec;
+using namespace facebook::velox::exec::test;
+using namespace facebook::velox::functions::aggregate::test;
+
+namespace facebook::velox::functions::aggregate::sparksql::test {
+
+namespace {
+
+class ModeAggregateTest : public AggregationTestBase {
+ protected:
+  void SetUp() override {
+    AggregationTestBase::SetUp();
+    registerAggregateFunctions("");
+  }
+
+  void testModeWithDuck(const VectorPtr& vector1, const VectorPtr& vector2) {
+    ASSERT_EQ(vector1->size(), vector2->size());
+
+    auto num = vector1->size();
+    auto reverseIndices = makeIndicesInReverse(num);
+
+    auto vectors = makeRowVector(
+        {vector1,
+         vector2,
+         wrapInDictionary(reverseIndices, num, vector1),
+         wrapInDictionary(reverseIndices, num, vector2)});
+
+    createDuckDbTable({vectors});
+    testAggregations(
+        {vectors},
+        {"c0"},
+        {"mode(c1)"},
+        "SELECT c0, mode(c1) FROM tmp GROUP BY c0");
+
+    testAggregations(
+        {vectors},
+        {"c0"},
+        {"mode(c3)"},
+        "SELECT c0, mode(c3) FROM tmp GROUP BY c0");
+
+    testAggregations(
+        {vectors},
+        {"c2"},
+        {"mode(c1)"},
+        "SELECT c2, mode(c1) FROM tmp GROUP BY c2");
+
+    testAggregations(
+        {vectors},
+        {"c2"},
+        {"mode(c3)"},
+        "SELECT c2, mode(c3) FROM tmp GROUP BY c2");
+  }
+
+  void testGlobalModeWithDuck(const VectorPtr& vector) {
+    auto num = vector->size();
+    auto reverseIndices = makeIndicesInReverse(num);
+
+    auto vectors =
+        makeRowVector({vector, wrapInDictionary(reverseIndices, num, vector)});
+
+    createDuckDbTable({vectors});
+    testAggregations({vectors}, {}, {"mode(c0)"}, "SELECT mode(c0) FROM tmp");
+
+    testAggregations({vectors}, {}, {"mode(c1)"}, "SELECT mode(c1) FROM tmp");
+  }
+
+  void testMode(
+      const std::string& expression,
+      const std::vector<std::string>& groupKeys,
+      const VectorPtr& vector1,
+      const VectorPtr& vector2,
+      const RowVectorPtr& expected) {
+    auto vectors = makeRowVector({vector1, vector2});
+    testAggregations({vectors}, groupKeys, {expression}, {expected});
+  }
```

Comment:
Could you clarify a bit when we will verify the results with DuckDB and when with the expected result? Thanks.

Reply:
Yes, there are 3 cases we don't verify with DuckDB :
1. complex types
2. bool, date, timestamp, looks Duckdb generates some unexpected results
3. there are cases we want to assert with explicit results like test cases: `globalNaNs`, `allNulls`, `globalEmpty`

Reply:
> bool, date, timestamp, looks Duckdb generates some different results

Is that due to semantic difference between Spark and DuckDB? This is tool to test DuckDB query online: https://shell.duckdb.org/.

Reply:
Thanks for the information.
For bool it complains bool can't be convert to int8_t.
`C++ exception with description "Invalid Input Error: Could not convert string 'true' to INT8" thrown in the test body.`
For timestamp and date, I think it's related to precision.
```
4 extra rows, 4 missing rows
4 of extra rows:
        null | "1970-01-01T00:00:00.017123456"
        0 | "1970-01-01T00:00:01.017123456"
        1 | "1970-01-01T00:00:01.017123456"
        2 | "1970-01-01T00:00:00.017123456"

4 of missing rows:
        null | "1970-01-01T00:00:00.017123000"
        0 | "1970-01-01T00:00:01.017123000"
        1 | "1970-01-01T00:00:01.017123000"
        2 | "1970-01-01T00:00:00.017123000"

Note: DuckDB only supports timestamps of millisecond precision. If this test involves timestamp inputs, please make sure you use the right precision.

```